### PR TITLE
Wait sensor async

### PR DESCRIPTION
**Pull request by collinmcnulty**
There is an example of a deferrable sensor that waits for one hour [in the documentation](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/deferring.html#writing-deferrable-operators), but there has not yet been an implementation of such a sensor in Airflow. This seems like a such a basic use case that it would make sense to include. There is the TimeDeltaSensor, but this sensor is based on the data interval of the DAG run, not the start time of the task.

I have observed some Airflow users using BashOperators that sleep for a specified time for want of this sensor, which would be strictly better, especially when run as deferrable.

---

**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: tests/sensors/test_time_delta.py [hunk 1]
```diff
@@ -71,3 +71,17 @@ def test_timedelta_sensor(self, defer_mock, should_defer):
...
         else:
             defer_mock.assert_not_called()
+
+    @pytest.mark.parametrize(
+        "should_defer",
+        [False, True],
+    )
+    @mock.patch("airflow.models.baseoperator.BaseOperator.defer")
+    def test_wait_sensor(self, defer_mock, should_defer):
+        delta = timedelta(seconds=30)
```
Comment (RNHTTR) [inline tests/sensors/test_time_delta.py]: I wonder if you can patch `asyncio.sleep` so the test doesn't have to wait at all

File: tests/sensors/test_time_delta.py [hunk 2]
```diff
@@ -71,3 +71,17 @@ def test_timedelta_sensor(self, defer_mock, should_defer):
...
+        op = WaitSensor(task_id="wait_sensor_check", delta=delta, dag=self.dag, deferrable=should_defer)
+        op.execute({})
+        if should_defer:
+            defer_mock.assert_called_once()
```
Comment (RNHTTR) [inline tests/sensors/test_time_delta.py]: Can you use `assert_called_once_with(...)`?

File: airflow/sensors/time_delta.py [hunk 3]
```diff
@@ -89,3 +92,38 @@ def execute(self, context: Context) -> bool | NoReturn:
...
+
+
+class WaitSensor(BaseSensorOperator):
+    """
+    A sensor that waits a specified period of time before completing.
+
+    This differs from TimeDeltaSensor because the time to wait is measured from the start of the task, not
+    the data_interval_end of the DAG run.
+
+    :param delta: time length to wait after the task starts before succeeding.
```
Comment (RNHTTR) [inline airflow/sensors/time_delta.py]: What about `time_to_wait` instead of `delta`?

**Review by RNHTTR — CHANGES_REQUESTED** [review]

File: tests/sensors/test_time_delta.py (hunk 2 above)
Comment (collinmcnulty) [reply tests/sensors/test_time_delta.py]: Trigger class isn't set up for this unfortunately.

```
FAILED tests/sensors/test_time_delta.py::TestTimeDeltaSensorAsync::test_wait_sensor[True] - AssertionError: expected call not found.
Expected: defer(trigger=<airflow.triggers.temporal.DateTimeTrigger moment=2024-08-01 00:00:30+00:00, end_from_trigger=True>, method_name='execute_complete')
  Actual: defer(trigger=<airflow.triggers.temporal.DateTimeTrigger moment=2024-08-01 00:00:30+00:00, end_from_trigger=True>, method_name='execute_complete')
```

File: airflow/sensors/time_delta.py (hunk 3 above)
Comment (collinmcnulty) [reply airflow/sensors/time_delta.py]: will do

File: airflow/sensors/time_delta.py [hunk 4]
```diff
@@ -89,3 +92,37 @@ def execute(self, context: Context) -> bool | NoReturn:
...
+    def __init__(
+        self,
+        time_to_wait: timedelta | int,
+        deferrable: bool = conf.getboolean("operators", "default_deferrable", fallback=False),
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.deferrable = deferrable
+        if isinstance(time_to_wait, int):
+            self.delta = timedelta(minutes=time_to_wait)
```
Comment (RNHTTR) [inline airflow/sensors/time_delta.py]:
```suggestion
            self.time_to_wait = timedelta(minutes=time_to_wait)
```

File: airflow/sensors/time_delta.py [hunk 5]
```diff
@@ -89,3 +92,37 @@ def execute(self, context: Context) -> bool | NoReturn:
...
+        else:
+            self.delta = time_to_wait
+
+    def execute(self, context: Context) -> None:
+        if self.deferrable:
+            self.defer(
+                trigger=TimeDeltaTrigger(self.delta, end_from_trigger=True),
```
Comment (RNHTTR) [inline airflow/sensors/time_delta.py]:
```suggestion
                trigger=TimeDeltaTrigger(self.time_to_wait, end_from_trigger=True),
```

File: airflow/sensors/time_delta.py [hunk 6]
```diff
@@ -89,3 +92,37 @@ def execute(self, context: Context) -> bool | NoReturn:
...
+        time_to_wait: timedelta | int,
+        deferrable: bool = conf.getboolean("operators", "default_deferrable", fallback=False),
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.deferrable = deferrable
+        if isinstance(time_to_wait, int):
+            self.delta = timedelta(minutes=time_to_wait)
+        else:
+            self.delta = time_to_wait
```
Comment (RNHTTR) [inline airflow/sensors/time_delta.py]:
```suggestion
            self.time_to_wait = time_to_wait
```

File: airflow/sensors/time_delta.py [hunk 7]
```diff
@@ -89,3 +92,37 @@ def execute(self, context: Context) -> bool | NoReturn:
...
+
+    def execute(self, context: Context) -> None:
+        if self.deferrable:
+            self.defer(
+                trigger=TimeDeltaTrigger(self.delta, end_from_trigger=True),
+                method_name="execute_complete",
+            )
+        else:
+            sleep(int(self.delta.total_seconds()))
```
Comment (RNHTTR) [inline airflow/sensors/time_delta.py]:
```suggestion
            sleep(int(self.time_to_wait.total_seconds()))
```

**Review by RNHTTR — APPROVED** [review]

**Review by vincbeck — APPROVED** [review]

**Pull request merged**